### PR TITLE
Implement buffered/delayed commands

### PR DESCRIPTION
**Pull request by Abyss777**
- If device is not online commands put in by device queues
- When device becomes online again traccar send all queued commands.

Queues are in memory only, lost if server restarted.

More comments in code

### CODE REVIEW

File: src/org/traccar/api/resource/CommandResource.java [hunk 1]
```diff
@@ -62,15 +62,19 @@ public Response send(Command entity) throws Exception {
...
-            Context.getCommandsManager().sendCommand(id, deviceId);
+            sent = Context.getCommandsManager().sendCommand(id, deviceId);
         } else {
             Context.getPermissionsManager().checkLimitCommands(getUserId());
             Context.getPermissionsManager().checkDevice(getUserId(), deviceId);
-            Context.getCommandsManager().sendCommand(entity);
+            sent = Context.getCommandsManager().sendCommand(entity, deviceId);
+        }
+        if (!sent) {
+            return Response.accepted(entity).build();
```
Comment (Abyss777) [inline src/org/traccar/api/resource/CommandResource.java]: I decided to notify API client that command was queued. I think code "202" - "Accepted" fits.

File: src/org/traccar/database/CommandsManager.java [hunk 2]
```diff
@@ -31,26 +35,22 @@
...
+    public boolean sendCommand(long commandId, long deviceId) throws Exception {
+        return sendCommand(getById(commandId), deviceId);
     }
 
-    public void sendCommand(long commandId, long deviceId) throws Exception {
-        sendCommand(getById(commandId), deviceId, false);
-    }
-
-    public void sendCommand(Command command, long deviceId, boolean fallbackToText) throws Exception {
+    public boolean sendCommand(Command command, long deviceId) throws Exception {
```
Comment (Abyss777) [inline src/org/traccar/database/CommandsManager.java]: return `true` if command sent, `false` if queued

File: src/org/traccar/database/CommandsManager.java [hunk 3]
```diff
@@ -71,32 +71,26 @@ public void sendCommand(Command command, long deviceId, boolean fallbackToText)
             if (activeDevice != null) {
                 activeDevice.sendCommand(command);
             } else {
-                if (fallbackToText) {
-                    command.setTextChannel(true);
-                    sendCommand(command, deviceId, false);
-                } else {
-                    throw new RuntimeException("Device is not online");
-                }
+                sent = !getDeviceQueue(deviceId).add(command);
```
Comment (Abyss777) [inline src/org/traccar/database/CommandsManager.java]: `add` always return true if not exception

File: src/org/traccar/database/CommandsManager.java [hunk 4]
```diff
@@ -71,32 +71,26 @@ public void sendCommand(Command command, long deviceId, boolean fallbackToText)
...
-                if (lastPosition != null) {
-                    BaseProtocol protocol = Context.getServerManager().getProtocol(lastPosition.getProtocol());
-                    if (protocol.getSupportedTextCommands().contains(command.getType())
-                            || online && protocol.getSupportedDataCommands().contains(command.getType())) {
-                        result.add(commandId);
-                    }
-                } else if (command.getType().equals(Command.TYPE_CUSTOM)) {
+            if (lastPosition != null) {
+                BaseProtocol protocol = Context.getServerManager().getProtocol(lastPosition.getProtocol());
+                if ((command.getTextChannel()
```
Comment (Abyss777) [inline src/org/traccar/database/CommandsManager.java]: This construction might looks unclear, can unwrap ternary operator.

File: src/org/traccar/database/CommandsManager.java [hunk 5]
```diff
@@ -133,4 +127,22 @@ public void sendCommand(Command command, long deviceId, boolean fallbackToText)
...
+        }
+        return deviceQueues.get(deviceId);
+    }
+
+    public void sendQueuedCommands(ActiveDevice activeDevice) {
+        Queue<Command> deviceQueue = deviceQueues.get(activeDevice.getDeviceId());
+        if (deviceQueue != null) {
+            Command command = deviceQueue.poll();
+            while (command != null) {
+                activeDevice.sendCommand(command);
```
Comment (Abyss777) [inline src/org/traccar/database/CommandsManager.java]: If exception happen here, command will be lost and rest of queue will be sent when device goes online again.

File: src/org/traccar/database/ConnectionManager.java [hunk 6]
```diff
@@ -122,6 +124,7 @@ public void updateDevice(final long deviceId, String status, Date time) {
                 public void run(Timeout timeout) throws Exception {
                     if (!timeout.isCancelled()) {
                         updateDevice(deviceId, Device.STATUS_UNKNOWN, null);
+                        activeDevices.remove(deviceId);
```
Comment (Abyss777) [inline src/org/traccar/database/ConnectionManager.java]: Assuming device is not active when its state unknown, to not send commands in emptiness.

**Comment by tsmgeek** [comment]
Is there to be any consideration if commands could have a time limit or if a user was to send the same command multiple times.

It may be an idea to put these commands in a DB to persist a restart, there are asset trackers which may have 24h intervals, a restart will loose any pending commands for these.

File: src/org/traccar/database/CommandsManager.java [hunk 7]
```diff
@@ -31,26 +35,22 @@
 
 public class CommandsManager  extends ExtendedObjectManager<Command> {
 
-    private boolean fallbackToText;
+    private final Map<Long, Queue<Command>> deviceQueues = new ConcurrentHashMap<>();
```
Comment (tananaev) [inline src/org/traccar/database/CommandsManager.java]: I know queue sounds like a right choice here, but really we can as well use regular list here. As soon as device is online we send all commands anyway.

File: src/org/traccar/database/CommandsManager.java [hunk 8]
```diff
@@ -133,4 +127,22 @@ public void sendCommand(Command command, long deviceId, boolean fallbackToText)
...
+            deviceQueues.put(deviceId, new ConcurrentLinkedQueue<Command>());
+        }
+        return deviceQueues.get(deviceId);
+    }
+
+    public void sendQueuedCommands(ActiveDevice activeDevice) {
+        Queue<Command> deviceQueue = deviceQueues.get(activeDevice.getDeviceId());
+        if (deviceQueue != null) {
+            Command command = deviceQueue.poll();
+            while (command != null) {
```
Comment (tananaev) [inline src/org/traccar/database/CommandsManager.java]: Regular for each loop should work here as well. And is simpler.

**Comment by tananaev** [comment]
@tsmgeek, it has been considered and for now we decided to go with simpler solution and in-memory buffer.

**Comment by Abyss777** [comment]
You are right. `ArrayList` is enough.

File: src/org/traccar/database/CommandsManager.java (hunk 4 above)
Comment (tananaev) [reply src/org/traccar/database/CommandsManager.java]: Just noticed this one. Please unwrap it.

**Comment by Abyss777** [comment]
Done

**Comment by tananaev** [comment]
I think I just found a bug. You have replaced queue poll with a for each loop, but now you don't clear the buffer.

**Comment by Abyss777** [comment]
Oh, sorry.
But now I see the problem, if we clear list after loop, it might cause infinite resend whole list in case exception. If we are going to remove entities in loop, we have to do it with help of `iterator` it will not look simple as we expected.

**Comment by tananaev** [comment]
You are right. Let's revert back to the queue.

**Comment by Abyss777** [comment]
Reverted.

**Comment by tsmgeek** [comment]
I'm just going to put a side case in, device is listed as connected but has just disappeared ie network just failed but socket has yet to close.
Would the message be queued or lost, I guess there is no confirmation of reception by device, or will it work on if an ACK was received from TCP packet, UDP another issue.

**Comment by tananaev** [comment]
Merged, thanks.

**Comment by tsmgeek** [comment]
Will the API return a flag to say the command was queued?

**Comment by Abyss777** [comment]
@tsmgeek 
It will return 202 code

**Pull request merged**